### DISCUSSION TITLE
Casting to user-defined type (postgresql enum) in SqlAlchemy core select query

### DISCUSSION THREAD

**Discussion by hozn** [sqlalchemy/sqlalchemy#8246] [Usage Questions] [answered]
I'm using SqlAlchemy 1.4.37 and Postgres 14.2

I'm using a Postgres ENUM on one of my tables, defined like:
```python
# The ENUM type is defined in DDL, like:
#.  create type custom_identifier_type AS ENUM ('type1', 'type2');

class EventModel(Base):
    iden = Column(UUID(as_uuid=True), primary_key=True,)
    identifier_type = Column(ENUM(name="custom_identifier_type", create_type=False))
    # snip
```

This works fine when inserting objects using the ORM, but I'm doing a insert-from-select where the selected table has the identifier_type as a varchar. 

I'm struggling with figuring out how to cast this using SA. In the select statement, I have tried to specify this using some form of literal -- naively:

```python
# src_table.c.identifier_type is VARCHAR (not `custom_identifier_type`)
insert(EventModel.__table__).\
  from_select(
   [EventModel.iden, EventModel.identifier_type], 
    sa.select(src_table.c.guid, sa.cast(src_table.c.identifier_type, sa.literal_column("custom_identifier_type"))
)
```

But I get errors in the form:
```
TypeError: Object BindParameter('%(11069711904 param)s', 'custom_identifier_type', type_=Unicode()) associated with '.type' attribute is not a TypeEngine class or object
```

Obviously I'm doing this wrong.  I was looking at custom types in SA, but that seemed overkill for simply wanting to put an explicit cast in here (also because this is already an ENUM, I assume it's more complicated than a simple pass-through).

Thanks in advance for any help!

**Comment by zzzeek** [comment, marked as answer]
cast places the expression first, type second.  also the "type" needs to be TypeEngine, not Column

```
cast("my_expression", mycolumn.type)
```

see docs at https://docs.sqlalchemy.org/en/14/core/sqlelement.html?highlight=cast#sqlalchemy.sql.expression.cast

**Reply by hozn** [reply]
Thanks for the reply.  (I typo'd an identifier in the original, so I updated for clarity.). Yes, I was attempting to generate an expression of the form `CAST(src_table.identifier_type, custom_identifier_type)` in the SELECT statement.

Trying to use the `.type` attribute of my target column doesn't cast it to my `custom_identifier_type` enum (it believes it to be VARCHAR, I guess):

```python
>> print(sa.select(src_table.c.guid, sa.cast(src_table.c.identifier_type, EventModel.identifier_type.type)))
SELECT src_table.guid, CAST(src_table.identifier_type AS VARCHAR) AS identifier_type 
FROM src_table
```
(I hoped I would get)
```sql
SELECT src_table.guid, CAST(src_table.identifier_type AS custom_identifier_type) AS identifier_type 
FROM src_table
```

**Reply by hozn** [reply]
As this problem is really particular to the insert/select combination I'm trying to (marshaling data between fdw tables and local), is there a way to inject a custom (literal?) expression here without fully defining a custom type?

**Reply by zzzeek** [reply]
you are printing the statement using the default compiler.  you [need to use the correct compiler to see the PostgreSQL specific form](https://docs.sqlalchemy.org/en/14/faq/sqlexpressions.html#stringifying-for-specific-databases)


```py
from sqlalchemy import cast
from sqlalchemy import Column
from sqlalchemy import insert
from sqlalchemy import select
from sqlalchemy import String
from sqlalchemy import Table
from sqlalchemy.dialects import postgresql
from sqlalchemy.dialects.postgresql import ENUM
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import declarative_base

Base = declarative_base()


class EventModel(Base):
    __tablename__ = "eventmodel"
    iden = Column(
        UUID(as_uuid=True),
        primary_key=True,
    )
    identifier_type = Column(
        ENUM(name="custom_identifier_type", create_type=False)
    )


src_table = Table(
    "src_table",
    Base.metadata,
    Column("guid", UUID(as_uuid=True)),
    Column("identifier_type", String),
)

stmt = insert(EventModel.__table__).from_select(
    [EventModel.iden, EventModel.identifier_type],
    select(
        src_table.c.guid,
        cast(src_table.c.identifier_type, EventModel.identifier_type.type),
    ),
)
print(stmt.compile(dialect=postgresql.dialect()))

```

output shows the correct statement:

```sql
INSERT INTO eventmodel (iden, identifier_type) SELECT src_table.guid, CAST(src_table.identifier_type AS custom_identifier_type) AS identifier_type 
FROM src_table
```

**Reply by hozn** [reply]
Thank you!  I should have considered this.  I greatly appreciate the detailed (and so immediate!) reply.
